Refactors how legs are displayed so they no longer appear above one-another when looking EAST or WEST

GoldenAlpharex opened the pull request:
## About The Pull Request
So, for ***over 5 years***, left legs have been displaying over right legs. Never noticed it? Don't blame you.
Here's a nice picture provided by https://github.com/tgstation/tgstation/issues/20603, that clearly displays the issue that was happening:
![image](https://user-images.githubusercontent.com/58045821/166085434-f942be98-d173-4166-b850-fc1d1d86b633.png)

It still happened to this day. Don't trust me? See for yourself:
![image](https://user-images.githubusercontent.com/58045821/166086180-22cf68dc-f500-4943-89a9-f9d0c10de8c0.png)
![image](https://user-images.githubusercontent.com/58045821/166086183-721e6c9e-dc49-4e22-afd2-95d4d9dd885a.png)
Notice how the two directions don't look the same? That's because the left leg is always displaying above the right one.

Obviously, that's no good, and I was like "oh, that's a rendering issue, so there's nothing I can do about it, it's an issue with BYOND".

Until it struck me.

*"What if we used a mask that would cut out the parts of the right leg, from the left leg, so that it doesn't actually look as if it's above it?"*

Here I am, after about 25 hours of work, 15 of which of *very* painful debugging due to BYOND's icon documentation sucking ass.

So, how does it work?

Basically, we create a mask of a left leg (that'll be explained later down the line), more specifically, a cutout of JUST the WEST dir of the left leg, with every other dir being just white squares. We then cache that mask in a static list on the right leg, so we don't generate it every single time, as that can be expensive. All that happens in `update_body_parts()`, where I've made it so legs are handled separately, to avoid having to generate limb icons twice in a row, due to it being expensive. In that, when we `generate_limb_icon()` a right leg, we apply the proper left leg mask if necessary.

Now, why masking the right leg, if the issue was the left leg?
Because, see, when you actually amputated someone, and gave them a leg again, it would end up being that new leg that would be displayed below the other leg. So I fixed that, by making it so that bodyparts would be sorted correctly, before the end of `update_body_parts()`. Which means that right legs ended up displaying above left legs, which meant that I had to change everything I had written to work on right legs rather than left legs.

I spent so much time looking up BYOND documentation for MapColors() and filters and all icon and image vars and procs, I decided to make a helper proc called `generate_icon_alpha_mask()`, because honestly it would've saved me at least half a day of pure code debugging if I had it before working on this refactor.

I tried to put as much documentation down as I could, because this shit messes with your brain if you spend too long looking at it. icon and image are two truly awful classes to work with, and I don't look forward to messing with them more in the future.

Anyway. It's nice, because it requires no other effort from anyone, no matter what the shape of the leg is actually like. It's all handled dynamically, and only one per type of leg, meaning that it's not actually too expensive either, which is *very* nice. Especially since it's very downstreams-friendly from being done this way.

Just look, both legs are now looking the same now:

![image](https://user-images.githubusercontent.com/58045821/166086019-9058df83-efe3-44de-8417-153f8bc33c57.png)
![image](https://user-images.githubusercontent.com/58045821/166086020-2afe22e2-b243-4d52-a0a1-fa89e89aff44.png)

## Why It's Good For The Game
It fixes https://github.com/tgstation/tgstation/issues/20603, an issue that has been around for over half a decade, as well as probably many more issues that I just didn't bother sifting through.

Plus, it just looks *so* much better. Look at this:
<details><summary>It works with Ethereal limbs</summary>

![image](https://user-images.githubusercontent.com/58045821/166085934-d0c53513-86af-4fca-a98c-c7b947214dca.png)
![image](https://user-images.githubusercontent.com/58045821/166085936-f5a76ca7-6910-4a72-9a5b-c27b4583b0a4.png)

</details>

<details><summary>It works with Moth legs</summary>

![image](https://user-images.githubusercontent.com/58045821/166085958-978b0cc5-7e68-4425-9e69-7d664d4171c6.png)
![image](https://user-images.githubusercontent.com/58045821/166085961-fc5be59f-9750-4d55-b390-ba61396688f8.png)

</details>

<details><summary>It even works with digitigrade legs!</summary>

![image](https://user-images.githubusercontent.com/58045821/166085981-b92b4fa7-27a6-48ae-8b54-0c159210f9da.png)
![image](https://user-images.githubusercontent.com/58045821/166085982-d153cb55-6658-49e4-8399-7d1280cbb087.png)

</details>

## Changelog

:cl: GoldenAlpharex
refactor: Refactored how legs are rendered, using dynamically-generated masks to avoid left legs always displaying over right ones, or right legs above left ones.
/:cl:

GoldenAlpharex committed:
Adds a file for the mask for the left legs

GoldenAlpharex committed:
Added the right_leg_mask_cache on left legs

GoldenAlpharex committed:
It works, it's not pretty, but it works.

The left leg still has a weird overlay bug on the right leg, but it's not entirely above it anymore, at least!

GoldenAlpharex committed:
This sucked like at least eight hours of my life just waiting for compilation and debugging, and then asking questions

I'm so tired, I'll fix the comments and stuff later.

GoldenAlpharex committed:
IT WORKS!!!!!

GoldenAlpharex committed:
Adds some new cmp for sorting bodyparts

GoldenAlpharex committed:
Made the mask actually applied to the right leg instead of the left one

GoldenAlpharex committed:
Removed some commented-out code

GoldenAlpharex committed:
Adds some documentation to generate_left_leg_mask

GoldenAlpharex committed:
Removes an fcopy_rsc() because turns out it's not even needed!

GoldenAlpharex committed:
Fixes two typos in comments

GoldenAlpharex committed:
Properly documents generate_masked_right_leg()

GoldenAlpharex committed:
Indicates the return type of generate_left_leg_mask()

GoldenAlpharex committed:
Adds a new helper, generate_icon_alpha_mask, for generating alpha masks for icons

GoldenAlpharex committed:
Replaces the MapColors in generate_left_leg_mask(), replaced it with the new icon helper

GoldenAlpharex committed:
Fixed some documentation not including the possibility of returning null

GoldenAlpharex committed:
Added image_dir to generate_masked_right_leg, so dropped legs don't look all weird anymore

Kapu1178 commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -237,27 +237,66 @@
 
 	//GENERATE NEW LIMBS
 	var/list/new_limbs = list()
+	bodyparts = sort_list(bodyparts, /proc/cmp_bodypart_by_body_part_asc)
```
Is there a reason this needs to be sorted every time?

GoldenAlpharex replied:
I couldn't think of any other way to know that the order of the bodyparts changed, and I figured it likely isn't going to take too many operations considering they'll be in order most of the time either way. If you have a better idea on how to do it, I'm all ears.

Kapu1178 replied:
Eh should be fine. God damn fuckin' byond demons making it so i cant make the bodyparts list associative 😭

GoldenAlpharex replied:
Maybe eventually we can figure a way, it would be *so* much easier, hahaha.

Kapu1178 commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -237,27 +237,66 @@
 
+
+	// Here we handle legs differently, because legs are a mess due to layering code. So we got to process the left leg first. Thanks BYOND.
+	var/list/left_leg_icon // yes it's actually a list, bet you didn't expect that, now did you?
+	var/leg_need_redrawn = FALSE
+	if(left_leg)
+		if(!(icon_render_keys[left_leg.body_zone] == old_left_leg_key))
+			left_leg_icon = left_leg.get_limb_icon()
+			limb_icon_cache[icon_render_keys[left_leg.body_zone]] = left_leg_icon
+			leg_need_redrawn = TRUE
+		else
+			left_leg_icon = limb_icon_cache[icon_render_keys[left_leg.body_zone]]
+
+	if(right_leg)
+		var/old_right_leg_key = icon_render_keys?[right_leg.body_zone]
+		var/left_leg_mask_key = left_leg?.generate_mask_key().Join()
+		right_leg.left_leg_mask_key = left_leg ? left_leg_mask_key : null
+		if(left_leg_mask_key && !right_leg.left_leg_mask_cache[left_leg_mask_key] && left_leg_icon)
+			right_leg.left_leg_mask_cache[left_leg_mask_key] = generate_left_leg_mask(left_leg_icon[1], left_leg_mask_key)
+		icon_render_keys[right_leg.body_zone] = right_leg.is_husked ? right_leg.generate_husk_key().Join() : right_leg.generate_icon_key().Join()
+
+		if(!(icon_render_keys[right_leg.body_zone] == old_right_leg_key))
+			limb_icon_cache[icon_render_keys[right_leg.body_zone]] = right_leg.get_limb_icon()
+			leg_need_redrawn = TRUE
```
line-by-line documentation pls.

This code is absurdly important and rather hot, and for ~~my~~ people who may need to read it's sanity, knowing exactly whats happening at any point in the proc is important.

GoldenAlpharex replied:
Yeah, I could add some more documentation here.

Kapu1178 commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -237,27 +237,66 @@
+	var/leg_need_redrawn = FALSE
+	if(left_leg)
+		if(!(icon_render_keys[left_leg.body_zone] == old_left_leg_key))
+			left_leg_icon = left_leg.get_limb_icon()
```
Not really sure how comfortable I am having any `get_limb_icon()` calls above the "main" return line. Overall this proc is SIGNIFICANTLY harder to read and understand.

GoldenAlpharex replied:
I totally agree, I could probably just move that out to a separate proc, I hadn't done it yet just out of slight concern for performance, but honestly... I don't think it matters enough, yeah.

Kapu1178 commented:
Also, on top of all that schtuff, a semi-detailed guide on how to make masks for other legs would be appreciated for people who want to add more non-humanoid legs, be it here or on other servers using Kapulimbs.

GoldenAlpharex commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -335,3 +397,58 @@
+ *
+ * Returns the `/image` of the right leg that was masked, or `null` if the mask didn't exist.
+ */
+/obj/item/bodypart/r_leg/proc/generate_masked_right_leg(right_leg_icon_file, right_leg_icon_state, image_dir)
+	if(!left_leg_mask_key || !left_leg_mask_cache[left_leg_mask_key] || !right_leg_icon_file || !right_leg_icon_state)
```
I highly considered it, until remembering it can be null too. And since I wasn't confident with how it works, I just didn't do it just to not make it so people expect this to be something else.

Kapu1178 replied:
RETURN_TYPE() allows nulls too, its common behavior for a proc to either return 1 type or null.

GoldenAlpharex replied:
Makes sense, I'm just used to having to specify returning undefined in typescript so it makes me a little wary, and I somewhat wish I could specify it can return null, if possible.

Kapu1178 replied:
It's only purpose is to allow things like 
thing.return_list().Join()

I wish we had ACTUAL return types 😭

GoldenAlpharex commented:
> Also, on top of all that schtuff, a semi-detailed guide on how to make masks for other legs would be appreciated for people who want to add more non-humanoid legs, be it here or on other servers using Kapulimbs.

I tried to make my procs as clear as possible, including the helper proc that generates the mask for you. I'll specify why I use the ICON_OR afterwards though, you're right, I'd say that's pretty much the only place it's not clear what I'm doing in there.

Aside from that, I'd like to think I explained stuff quite a decent bit in my PR body.

GoldenAlpharex committed:
Added some more return types

GoldenAlpharex committed:
Removed a check that didn't need to be there

GoldenAlpharex committed:
Adds more documentation to generate_left_leg_mask() to explain the icon Blend()s, to make it easier to understand in the future

GoldenAlpharex committed:
Moved all the less-pretty code into its own proc, to improve readability in update_body_parts()

Kapu1178 commented:
I pray nobody needs to touch this code in the future, myself included.

optimumtact commented:
what's the impact on map tick

timothymtorres commented:
Amen to knocking out one of the bugs that is currently the 9th oldest issue on the repo.  It's probably older than 5 years and nobody bothered to open a bug report on it.  I could see this issue existing since when Exadv1 made the game since they didn't have to worry about other sprites besides humans.

Out of curiosity, why does this only affect legs though?  Why couldn't you use the code that displays arms in the correct dir?

GoldenAlpharex commented:
> what's the impact on map tick

Ideally, nothing noticeable, considering this should be done very rarely within the span of a round, and mostly done at the start of said round.

That being said, I don't quite know how to measure map tick, and I'd probably need the help of someone more knowledgeable than me on the matter to really provide a worthy answer to this question.

GoldenAlpharex commented:
> I pray nobody needs to touch this code in the future, myself included.
> 
> You did test to make sure the caching works properly and doesn't redraw bodyparts more than it needs to, right?

I'll do some more testing to make sure, but ideally it shouldn't, the caching logic is basically the same, there's just more steps beforehand to update the mask key on the right leg, which is something that the right leg's icon_key depends on.

GoldenAlpharex commented:
> Amen to knocking out one of the bugs that is currently the 9th oldest issue on the repo. It's probably older than 5 years and nobody bothered to open a bug report on it. I could see this issue existing since when Exadv1 made the game since they didn't have to worry about other sprites besides humans.

I didn't realize that it was actually the 9th-oldest issue, hahaha. But yeah, this probably has been around for even longer, yeah.

> Out of curiosity, why does this only affect legs though? Why couldn't you use the code that displays arms in the correct dir?

Basically, it only applies to legs because there's different code for the arms:
Arms are rendered in two parts, that being the arm itself, and the hand. On the EAST direction, there's simply no sprite for the left arm, and on the WEST direction, there's no sprite for the right arm. There's only sprites for a teeny tiny bit of the hand, on these directions.

Now, if I wanted to really be exhaustive, I could come back to this in the future and do mostly the same for arms, which would allow different body shapes to mask the arms differently. That being said, the code for that would likely be *even* uglier than what there is for the legs right now, because I'd likely have to take undershirts into account (unless they display above hands, which would be a blessing in this case), and the masking would have to depend on a mix of: the opposing hand sprite, the opposing arm sprite and the chest sprite.

Which is a lot of spaghetti, and considering how it's already very functional for arms right now, and that you wouldn't really notice it unless you were to be a skeleton/plasmaman, I think the cost of it would outweigh the function, in this case.

GoldenAlpharex commented:
> I pray nobody needs to touch this code in the future, myself included.
> 
> You did test to make sure the caching works properly and doesn't redraw bodyparts more than it needs to, right?

I tested it again, and no, it doesn't redraw them more than it should!

LemonInTheDark commented on code/__HELPERS/icons.dm:
```diff
@@ -897,6 +897,37 @@ world
+	// colors, and then the fourth group of number arguments is actually the alpha values of
+	// each of the original colors, which we multiply by 255 and subtract a value of 255 to the
+	// result for the matching pixels, while starting with a base color of white everywhere.
+	mask_icon.MapColors(0,0,0,0, 0,0,0,0, 0,0,0,0, 255,255,255,-255, 1,1,1,1)
```
why use 255 here? this should be 0 - 1
what do you actually want to happen?

GoldenAlpharex replied:
It's 255 because I want to ensure that it actually does not use the alpha value if it has one, and makes it 0%, pretty much.

255 ensures that as long as there's at least 1 of the value, it gets taken. This is how I can ensure that the colors are properly handled.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -237,27 +237,44 @@
 	for(var/obj/item/bodypart/limb as anything in bodyparts)
 		limb.update_limb(is_creating = update_limb_data) //Update limb actually doesn't do much, get_limb_icon is the cpu eater.
+
+		if(limb.body_zone == BODY_ZONE_R_LEG)
+			right_leg = limb
+			continue // Legs are handled separately
+
 		var/old_key = icon_render_keys?[limb.body_zone] //Checks the mob's icon render key list for the bodypart
 		icon_render_keys[limb.body_zone] = (limb.is_husked) ? limb.generate_husk_key().Join() : limb.generate_icon_key().Join() //Generates a key for the current bodypart
+
+		if(limb.body_zone == BODY_ZONE_L_LEG)
+			left_leg = limb
+			old_left_leg_key = old_key
+			continue // Legs are handled separately
```
why is this done like this? is the order important?

GoldenAlpharex replied:
Yes, it's to avoid regenerating a icon_key unnecessarily, because we have to do it later on for he right leg, once he left leg has been handled.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -237,27 +237,44 @@
 
 	//GENERATE NEW LIMBS
 	var/list/new_limbs = list()
+	bodyparts = sort_list(bodyparts, /proc/cmp_bodypart_by_body_part_asc)
```
If possible it would be better to sort this on insertion into the bodyparts list, rather then sorting each time you update overlays fully.

GoldenAlpharex replied:
It likely would be possible, yeah. I didn't really try that, truth be told.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +283,53 @@
+	var/list/left_leg_icon // yes it's actually a list, bet you didn't expect that, now did you?
+	var/legs_need_redrawn = FALSE
+	if(left_leg)
+		// We regenerate the look of the left legif it isn't already cached, we don't if not.
```
legif

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +283,53 @@
+	var/legs_need_redrawn = FALSE
+	if(left_leg)
+		// We regenerate the look of the left legif it isn't already cached, we don't if not.
+		if(!(icon_render_keys[left_leg.body_zone] == old_left_leg_key))
+			left_leg_icon = left_leg.get_limb_icon()
+			limb_icon_cache[icon_render_keys[left_leg.body_zone]] = left_leg_icon
+			legs_need_redrawn = TRUE
+		else
+			left_leg_icon = limb_icon_cache[icon_render_keys[left_leg.body_zone]]
```
```suggestion
		if(!(icon_render_keys[left_leg.body_zone] == old_left_leg_key))
			limb_icon_cache[icon_render_keys[left_leg.body_zone]] = left_leg.get_limb_icon()
			legs_need_redrawn = TRUE
		left_leg_icon = limb_icon_cache[icon_render_keys[left_leg.body_zone]]
```
Up  to you, personally I like this pattern more

GoldenAlpharex replied:
That looks good, will commit in the morning.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +283,53 @@
+		// We generate a new icon_render_key, which also takes into account the left_leg_mask_key so we cache the masked versions of the limbs too.
+		icon_render_keys[right_leg.body_zone] = right_leg.is_husked ? right_leg.generate_husk_key().Join() : right_leg.generate_icon_key().Join()
+
+		if(!(icon_render_keys[right_leg.body_zone] == old_right_leg_key))
+			limb_icon_cache[icon_render_keys[right_leg.body_zone]] = right_leg.get_limb_icon()
+			legs_need_redrawn = TRUE
```
Why does this exist?

GoldenAlpharex replied:
What do you mean? It's to update the icon of the right leg, I'm not sure what's wrong with that?

LemonInTheDark replied:
ah I see

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -335,3 +416,61 @@
+ */
+/obj/item/bodypart/r_leg/proc/generate_masked_right_leg(right_leg_icon_file, right_leg_icon_state, image_dir)
+	RETURN_TYPE(/image)
+	if(!left_leg_mask_key || !left_leg_mask_cache[left_leg_mask_key] || !right_leg_icon_file || !right_leg_icon_state)
```
would `left_leg_mask_cache[null]` be possible in the first place?

GoldenAlpharex replied:
If there's no left leg, yeah, it would be.

LemonInTheDark replied:
What would it return? also null?

GoldenAlpharex replied:
> What would it return? also null?

Oh, no, uh, I don't think that doing `left_leg_mask_cache[null]` would work, no. But that's why I did a key check first.

GoldenAlpharex replied:
> > What would it return? also null?
> 
> Oh, no, uh, I don't think that doing `left_leg_mask_cache[null]` would work, no. But that's why I did a key check first.

It does work, I'm dumb, I'll just shorten it.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -335,3 +416,61 @@
+		return
+
+	var/icon/right_leg_icon = icon(right_leg_icon_file, right_leg_icon_state)
+	right_leg_icon.Blend(left_leg_mask_cache[left_leg_mask_key], ICON_MULTIPLY)
+	return image(right_leg_icon, right_leg_icon_state, layer = -BODYPARTS_LAYER, dir = image_dir)
```
This hurts

GoldenAlpharex replied:
It does, but from ninja's own words, making an image out of an icon is actually very cheap, something something direct reference to rsc. I'd need to ask him again to be certain, though.

GoldenAlpharex committed:
Applied review comments + improved var name

Fixed legif in a comment and improved the look of a small part of the code in update_legs()

GoldenAlpharex committed:
Optimized the limbs sorting by making it only happen on attach_limb() rather than update_body_parts()

This way, it's not going to get called for no reason, so often.

Kapu1178 commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -365,6 +365,7 @@
 	if(can_be_disabled)
 		update_disabled()
 
+	new_limb_owner.bodyparts = sort_list(new_limb_owner.bodyparts, /proc/cmp_bodypart_by_body_part_asc)
```
Leave a comment here as to why this exists so someone doesnt fuck it up later

GoldenAlpharex replied:
True.

GoldenAlpharex committed:
Adds a comment to explain why we sort bodparts on Insert()

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +282,52 @@
+
+	if(right_leg)
+		var/old_right_leg_key = icon_render_keys?[right_leg.body_zone]
+		var/left_leg_mask_key = left_leg?.generate_mask_key().Join() // We generate a new mask key, to see if it changed.
+		right_leg.left_leg_mask_key = left_leg ? left_leg_mask_key : null
```
What is the point of this ternary? ?. will hand back null if the thing it's being called on is null

LemonInTheDark replied:
```suggestion
		right_leg.left_leg_mask_key = left_leg?.generate_mask_key().Join() // We generate a new mask key, to see if it changed.
```
could genuinely just do this

LemonInTheDark replied:
Also it's weird to call join with no arguments, opens you up to weird caching headaches, might wanna stick a "-" in there

GoldenAlpharex replied:
That's how it was done elsewhere, I just copied a proc and then removed some of the content.

Kapu1178 replied:
I am responsible for the `.Join()` meme. Key generation was originally a string until I was informed the horror that making a list and using `list.Join()` was faster than string concatenation. Since all of the key parts already were using `"-"`, I just slapped a `.Join()` there.

LemonInTheDark replied:
I hate coders

GoldenAlpharex replied:
I kept it as a separate var to make the following check a little shorter, I know it wasn't necessary, but I did it for readability, mostly.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +282,52 @@
+		var/old_right_leg_key = icon_render_keys?[right_leg.body_zone]
+		var/left_leg_mask_key = left_leg?.generate_mask_key().Join() // We generate a new mask key, to see if it changed.
+		right_leg.left_leg_mask_key = left_leg ? left_leg_mask_key : null
+		// We regenerate the left_leg_mask in case that it doesn't exist yet.
+		if(left_leg_mask_key && !right_leg.left_leg_mask_cache[left_leg_mask_key] && left_leg_icons)
+			right_leg.left_leg_mask_cache[left_leg_mask_key] = generate_left_leg_mask(left_leg_icons[1], left_leg_mask_key)
+		// We generate a new icon_render_key, which also takes into account the left_leg_mask_key so we cache the masked versions of the limbs too.
+		icon_render_keys[right_leg.body_zone] = right_leg.is_husked ? right_leg.generate_husk_key().Join() : right_leg.generate_icon_key().Join()
+
```
actually, I get the feeling this code would benifit from being made into a proc on the leg, to reduce the amount of thing.x you need to do
also could make the var names shorter, though I'm not sure how exactly you'd do that

GoldenAlpharex replied:
I personally prefer having slightly longer names than ambiguous ones.

Also, yes, I *could* do that, I didn't in this case because I wanted to leave it all in one place, so that it would be easy to find in case someone wanted to change behavior in there.

Kapu1178 replied:
I'd prefer `r_leg/l_leg` because it makes the code easier to read

GoldenAlpharex replied:
> I'd prefer `r_leg/l_leg` because it makes the code easier to read

I'd respectfully disagree, the difference between `r` and `l` can be tough to spot at a quick glance (especially when you've got attention problems), I dislike using abbreviations when I can write out the full word, it's easier to spot and to search for that way, in my personal opinion.

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -266,12 +282,52 @@
+		// We generate a new icon_render_key, which also takes into account the left_leg_mask_key so we cache the masked versions of the limbs too.
+		icon_render_keys[right_leg.body_zone] = right_leg.is_husked ? right_leg.generate_husk_key().Join() : right_leg.generate_icon_key().Join()
+
+		if(!(icon_render_keys[right_leg.body_zone] == old_right_leg_key))
```
```suggestion
		if(icon_render_keys[right_leg.body_zone] != old_right_leg_key)
```

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -302,6 +358,29 @@
+	RETURN_TYPE(/list)
+	. = list()
+	if(is_dimorphic)
+		. += "[limb_gender]-"
+	. += "[limb_id]"
+	. += "-[body_zone]"
+	for(var/obj/item/organ/external/external_organ as anything in external_organs)
+		if(!external_organ.can_draw_on_bodypart(owner))
+			continue
+		. += "-[external_organ.generate_icon_cache()]"
```
You could drop the -s here and just .Join("-") to achieve the same result with less headache

GoldenAlpharex replied:
I'm not quite sure what you mean here.

I'd have to change it for all of those procs, no?

LemonInTheDark replied:
see below

LemonInTheDark commented on code/modules/mob/living/carbon/carbon_update_icons.dm:
```diff
@@ -302,6 +358,29 @@
+ */
+/obj/item/bodypart/proc/generate_mask_key()
+	RETURN_TYPE(/list)
+	. = list()
```
rather then using . as a var, you should just make a new list called like, to_return or something.
If you're just going to treat . as a var all the time you'd better off not using . (unless your code is hideously hot, which this isn't)

GoldenAlpharex replied:
I'll be honest, I just copy-pasted this from Kapu's code, I don't mind changing it, I just kept it this way for consistency.

LemonInTheDark replied:
Kapu seems to be plannin on changing them. I'd like it if you could change it here tho

LemonInTheDark commented on code/modules/surgery/bodyparts/parts.dm:
```diff
@@ -361,6 +361,13 @@
 	can_be_disabled = TRUE
+	/// We store this here to generate our icon key more easily.
+	var/left_leg_mask_key
+	/// The associated list of all the left leg mask keys associated to their cached left leg masks.
+	/// It's static, so it's shared between all the left legs there is. Be careful.
+	/// Why? Both legs share the same layer for rendering, and since we don't want to do redraws on
+	/// each dir changes, we're doing it with a mask instead, which we cache for efficiency reasons.
+	var/static/list/left_leg_mask_cache = list()
```
any reason this can't just be a global list?

GoldenAlpharex replied:
It could be a global list, yeah, I just didn't really see a point in making it a global list rather than just a static list, since nothing else will access it anyway. Less chances of it being tampered with.

Kapu1178 replied:
GLOB.thing is slower to access than a static var isnt it

LemonInTheDark replied:
not in a way that matters for anything being done here

GoldenAlpharex committed:
Removes one useless check

GoldenAlpharex committed:
Removed another useless null check

GoldenAlpharex committed:
!(A == B) -> A != B

GoldenAlpharex committed:
Removed an useless variable and improved the join code a bit on generate_mask_key()
